api/src/main/java/io/druid/data/input/FirehoseFactory.java

```diff
@@ -36,7 +37,28 @@
    * If this method returns null, then any attempt to call hasMore(), nextRow(), commit() and close() on the return
    * value will throw a surprising NPE.   Throwing IOException on connection failure or runtime exception on
    * invalid configuration is preferred over returning null.
+   *
+   * @param parser             an input row parser
    */
-  public Firehose connect(T parser) throws IOException, ParseException;
+  @Deprecated
+  default Firehose connect(T parser) throws IOException, ParseException
+  {
+    return connect(parser, null);
+  }
 
+  /**
+   * Initialization method that connects up the fire hose.  If this method returns successfully it should be safe to
+   * call hasMore() on the returned Firehose (which might subsequently block).
+   * <p/>
+   * If this method returns null, then any attempt to call hasMore(), nextRow(), commit() and close() on the return
+   * value will throw a surprising NPE.   Throwing IOException on connection failure or runtime exception on
+   * invalid configuration is preferred over returning null.
+   * <p/>
+   * Some fire hoses like {@link PrefetchableTextFilesFirehoseFactory} may use a temporary
+   * directory to cache data in it.
+   *
+   * @param parser             an input row parser
+   * @param temporaryDirectory a directory where temporary files are stored
+   */
+  Firehose connect(T parser, File temporaryDirectory) throws IOException, ParseException;
```

Comment:
It's got to call the other method; the case we're trying to support is that someone already wrote an extension for Druid 0.10.0 with a custom FirehoseFactory, and then upgrades to Druid 0.10.1. The new Druid 0.10.1 code will call `connect(parser, temporaryDirectory)` and unless there is a default impl that calls `connect(parser)`, the user's extension will not work properly.

Generally, we want to keep extension compatibility for the extension points on http://druid.io/docs/latest/development/modules.html within a major version like 0.9.x, 0.10.x, etc (it's ok to break extensions from 0.9.x -> 0.10.x if there is a good reason).